src/components/ebay-combobox/component.js

```diff
@@ -179,17 +184,12 @@ export default {
         this._setupMakeup();
     },
 
-    onRender() {
-        if (typeof window !== 'undefined') {
-            this._cleanupMakeup();
-        }
-    },
-
```

Comment:
This will not correctly clean up when the component rerenders when this triggers. 
We do not need to touch the lifecycle methods. These were implemented by the marko team and can cause issues in certain server side rendering/rerendering cases.
We should call active descendant reset on click, like I recommended earlier.

Reply:
Can you tell me when it will not work? 
It seems it works fine. 

`_setupMakeup()` is called `onUpdate` and `onMount` and `_cleanupMakeup()` is called in it to clean before set things up again.

Reply:
The thing is the previous call is not being cleared correctly because the html does not exist anymore. 
we have this pattern in all our other components. Not going to deviate from that. 
lets add that reset when we open the dropdown. Is that causing issues or not working?

Reply:
But that pattern is what creates the #1782 bugs. And it may always be the cause of others.

## A better pattern to guarantee clean up and resource availability

When you have an resource that need to be available during the lifecycle of a object you usually have this pattern:

```
  create() {
    this.obj = new Resource();
  }

   destroy() {
        this.obj.destroy();
   }
``` 

### Recreate
**FOR EACH each setup, you want ONE (and ONLY ONE) cleanup of the resource** so the chain look like: 
```
setup -> cleanup -> setup -> cleanup -> ... -> cleanup
```


If every once in a while you need to destroy the resource (cleaning its side effects) before a new one is created, having **one**  place where the destruction and (re)building are done together in order is the **simplest way** to guarantee that the resource will be **always** available and each setup will have one cleanup: 
```
  create() {
    this.obj = new Resource();
  }

  recreate() { // Keep them together without gaps!
      if (this.obj) this.obj.destroy();
      this.obj = new Resource()
   }
   
   destroy() {
        if (this.obj) this.obj.destroy();
   }
``` 

Or in other more general sudocode:

```
  create() {
    setup();
  }

  recreate() {
      cleanup();
      setup();
  }
   
  destroy() {
       cleanup();
  }
``` 

Since the actions are synchronous a `cleanup()` and a `setup()` run in `recreate()` **without gaps**.  In Javascript you are guaranteed that the resource reference will always be a valid object. It also impossible to overcall or undercall the `setup` or the `cleanup`.*
To make things even more DRY, you can have a `cleanupIfNecessaryAndSetup()` instead so there are only 2 functions. 

```
cleanupIfNecessaryAndSetup() {
   if (this.resource) cleanup();
   setup(); // This will set this.resource
}
...
{
  create() {
    cleanupIfNecessaryAndSetup();
  }

  recreate() {
      cleanupIfNecessaryAndSetup();
  }
   
  destroy() {
       cleanup();
  }
}
``` 

A **problem** can like occur instead if you break down the `recreate` in 2 methods. You are not guaranteed anymore that there is no gap (ex: callers are asynchronous) or `setup` and `cleanup` are not overcalled or undercalled.

This is what I did here in this PR moving the `this._cleanupMakeup()` inside the `_setupMakeup()`. So the `_setupMakeup` became the `cleanupIfNecessaryAndSetup()` above:
```
onMount() {
     cleanupIfNecessaryAndSetup()
     ...
}

onUpdate() {
     cleanupIfNecessaryAndSetup()
     ...
}

onDestroy() {
     cleanup()
     ...
}
```

These are the [Marko lifecycle](https://markojs.com/docs/class-components/#lifecycle-events) flows:
```
First render
emit('create') → emit('input') → emit('render') → emit('mount')

New input
emit('input') → emit('render') → emit('update')

Internal state change
emit('render') → emit('update')

Destroy
emit("destroy");
```

And each of the last actions in the chain (`mount`, `update`, `update` and `destroy`)  has the right call in it.

Now if you keep this structure the code is robust against changes and it is impossible to have a bug related to the call timing.

Reply:
@cc-ebay without the full context of reading everything here:

The main issue is that although makeup follows a sane lifecycle, because it's an imperative API that touches the DOM in many places there is awkwardness here. Specifically, every time Marko renders the Makeup changes need to be _undone_ in order for Marko to properly complete it's update.

I believe it was mentioned in here somewhere that the ultimate goal is to recreate makeup or big portions of it as components to avoid this issue, but for now this is what we've got.

Reply:
Just to add, spoke to Dylan as well

We need to keep the onRender cleanup because by the time we hit onUpdate, the html has already been removed and there's nothing to cleanup. And we don't want to call cleanup twice. 
I think there's more to this that's causing this bug. It looks like the previous cleanup might not be resetting correctly or such. 
By adding that reset as I recommended, should resolve this issue while allowing the cleanup to happen properly as before.

Reply:
Ok. I understand that this could be a better idea to split the `reload` in 2. Then we are not guaranteed anymore that the resources will be always available.

Reply:
I also found the root cause of https://github.com/eBay/ebayui-core/issues/1782:
 https://github.com/eBay/ebayui-core/blob/5f1e819603d7776beb03083d374971a3b7e12456/src/components/ebay-combobox/component.js#L235
 
 Call again the `_setupMakeup()` so it is called twice on load. That setup 2 `activeDescendant` without cleaning in the middle.
 
Since the `_setupMakeup()` here is setting up 3 resources (`activeDescendant`, `expander`, `_floatingLabel`), it would be good to guard the actual call of the setup function for each resources checking for existence. Ex:
 ```
 if (this.activeDescendant) this.activeDescendant = createLinear(..
 ...
 if (!this.expander) this.expander = new Expander(...
 ...
 ```
 
 The `_floatingLabel` could also move in its own function (ex:  `_setupFloatingLabel`) and instead of calling the whole  `_setupMakeup()` on the `load` event, can call itself.